walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -24,19 +24,48 @@
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, 
                                      MeetingRequest request) {
-    // Get unavailable time for all required attendees in meeting request
     ArrayList<TimeRange> unavailable = new ArrayList<>();
+
+    // Get unavailable time for all required attendees in meeting request
     for (String attendee : request.getAttendees()) {
       ArrayList<TimeRange> meetingTimes = getMeetingTimes(events, attendee);
-      merge(unavailable, meetingTimes);
+      mergeTo(unavailable, meetingTimes);
     }
 
     // Get time ranges that are longer than the requested meeting duration
+    // for required attendees
     Collection<TimeRange> result = getAvailability(unavailable);
     result.removeIf( 
       timeRange -> timeRange.duration() < request.getDuration());
 
-    return result;
+    // Consider optional attendees
+    ArrayList<TimeRange> unavailableWithOptional = new ArrayList<>();
+    Collection<TimeRange> resultWithOptional = new ArrayList<>();
+    unavailableWithOptional.addAll(unavailable);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      // Merge unavailable time for all optional attendees in meeting request
+      for (String attendee : request.getOptionalAttendees()) {
+        ArrayList<TimeRange> meetingTimes = getMeetingTimes(events, attendee);
+        mergeTo(unavailableWithOptional, meetingTimes);
+      }
+
+      // Get time ranges that are longer than the requested meeting duration
+      // for all attendees
+      resultWithOptional = getAvailability(unavailableWithOptional);
+      resultWithOptional.removeIf( 
+        timeRange -> timeRange.duration() < request.getDuration());
+    } else {
+      // If there are no optional attendees, result does not change
+      resultWithOptional.addAll(result);
+    }
+
+    // If there are no mandatory attendees, treat optional as mandatory
+    if (request.getAttendees().isEmpty()) {
+      return resultWithOptional;
+    }
```

Comment:
It seems that you don't need this check?

Reply:
When there are no mandatory attendees, and the optional attendees happen to have conflict in their schedule, for example, notEnoughRoomForOptionalsOnly() in the test file
There should be an empty list returned, but my algorithm will return the whole day, since I return mandatory attendees availability (`result`) when `resultWithOptional` is empty. 
So I put this check there to treat optional as mandatory when there are no mandatory attendees.

Reply:
yes, you are right.